ddtrace/tracer/payload_v04.go

```diff
@@ -75,6 +75,21 @@ func newPayloadV04() *payloadV04 {
 func (p *payloadV04) push(t spanList) (stats payloadStats, err error) {
 	p.setTracerTags(t)
 	p.buf.Grow(t.Msgsize())
+
+	for _, span := range t {
+		if span != nil {
+			span.mu.RLock()
+		}
+	}
+
+	defer func() {
+		for _, span := range t {
+			if span != nil {
+				span.mu.RUnlock()
+			}
+		}
+	}()
```

Comment:
This shouldn't be needed. At this point the spans have been finished and flushed, so they are guarded by `s.finished`.

In the incident channel's thread I left a message where - after looking at logs - I think the issue might come from `Finish` because there is one place on it where a span isn't locked but its trace is locked. Specifically when the trace has spans, the current finished span is the first span of the trace, as it calls `t.setTraceTags`. If the latter condition isn't true, the same function is called on the first finished span of the trace.